torchcde/misc.py

```diff
@@ -136,15 +138,17 @@ def forward_fill(x):
     assert x.dim() >= 2
 
     if torch.isnan(x).any():
-        # Reshape to [-1, L]
-        L, C = x.size()[-2:]
-        x_shaped = x.transpose(-1, -2).reshape(-1, L)
+        # Fill index needs to be in the -1th dimensino
+        x_transposed = x.transpose(-1, fill_index)
+        # Reshape to [-1, length]
+        length = x_transposed.size(-1)
+        x_shaped = x_transposed.reshape(-1, length)
         # Fill the 2d tensor along the length dim
         mask = torch.isnan(x_shaped)
-        idx = torch.where(~mask, torch.arange(mask.shape[1]), 0)
+        idx = torch.where(~mask, torch.arange(mask.shape[1]).to(mask.device), 0)
         cummax_idx, _ = torch.cummax(idx, dim=1)
-        x_fill = x_shaped[torch.arange(cummax_idx.shape[0])[:, None], cummax_idx].reshape(-1, C, L)
-        # Map back to original size
-        x = x_fill.transpose(-1, -2).reshape(x.size())
+        x_filled = x_shaped[torch.arange(cummax_idx.shape[0])[:, None], cummax_idx].reshape(x_transposed.size())
+        # Back to normal dim
+        x = x_filled.transpose(-1, fill_index)
 
     return x
```

Comment:
Every comment previously made still applies.
Moreover it is once again clear that you have not actually tested this, as this code cannot run.

Try this instead:
```
def forward_fill(x, fill_index=-2):
    assert isinstance(x, torch.Tensor)
    assert x.dim() >= 2
    
    mask = torch.isnan(x)
    if mask.any():
        _, index = (~mask).cummax(dim=fill_index)
        x = x.gather(dim=fill_index, index=index)
        
    return x
```

I've not completely tested that but I think it's doing the right thing.

Bonus question for you: if using my version, what additional test should be added to ensure that things don't silently break in the future?

Reply:
So as not to tarnish my good name: It was tested, but turns out it fails for torch==1.6.0 but not torch==1.7.0. 

Neverthless, your code is far neater so I have replaced it. 

I have no idea regarding bonus questions, maybe testing cummax/gather to ensure they give the same outputs we expect now as the torch version changes?

Reply:
Hah, that's fair!

And yup you're nearly there with the bonus question.

My code relies on the indices returned by cummax always being the last occurence of the maximum. For example, as specified by the documentation for `torch.cummax` at time of writing, it is ambiguous whether `torch.tensor([2, 2]).cummax().indices` should give 0s deterministically, 1s deterministically, or to randomly choose between them according to whatever rule it wants.

In this case I found out that `cummax` happens to use the desired behaviour just by testing it, but to be certain there should be a test for this, in case this changes in some later version of PyTorch. The test should be quite comprehensive: try tensors with different devices, different numbers of dimensions, which dimension the cummax is taken down, sizes of dimension (inc. important special cases: 0, 1, "small"~5, values either side of magic numbers like 32, "large"~1000); and indeed combinations of the above. It only needs to test bool-type tensors though as that's what's used here.

Incidentally if you want a nice example about not making assumptions for this kind of thing (at least using PyTorch 1.6.0 on Ubuntu):
```
import torch

def f(dtype, size):
    g = torch.Generator().manual_seed(456)
    print(torch.randn(size, dtype=dtype, generator=g)[0].item())

f(torch.float32, 15)
f(torch.float64, 15)
```
Prints out the same thing both times. We might reasonably assume that the choice of dtype doesn't matter. But then:
```
f(torch.float32, 16)
f(torch.float64, 16)
```
prints different things! It turns out there's a change in behaviour once the size of the tensor exceeds 16. Some other algorithm is used for which the choice of dtype now _does_ matter.

In fact not only does it print two different things, but even for the same dtype it's not consistent with the `size=15` case. It turns out that all `size<15` are the same as each other, and all `size>=16` are the same as each other, and that in the former case `float32` and `float64` just _happen_ to be the same as well.

Moreover note that I'm using `randn`. Using `rand` gives different results again.

Reply:
I do not think it is as complicated as this. We have a boolean tensor and simply wish cummax to give us the argument of the closest previous True value along a given dimension. With your 2 example the only implementation detail is that `torch.tensor([2, 2]).cummax().indices` could return `[0, 1], or [0, 0]`. It could never give `[1, 1]` since at the 0th argument it cannot return a larger cummax argument. 

Therefore the only check is that it will give `[0, 1]` not `[0, 0]`, as we want the most recent location. Thus I believe the test 
```
indices_check = torch.tensor([False, False, True, True]).to(device).cummax(0).indices
desired_indices = torch.tensor([0, 1, 2, 3]).to(device)
assert torch.equal(indices_check, desired_indices)
```
is more than sufficient. It simply ensures that if we have two Trues (i.e. two observed values), it will return to us the argument of the most recent true.

Reply:
Cummax might do the correct thing on your particular test. But it might still do something else on a different input.

Anyway, relying on an implementation detail like this is pretty bad practice. Sorry, that's my bad - I shouldn't have suggested this as a solution in the first place.

Better (untested; going back to something like what you had):
```
    mask = torch.isnan(x)
    if mask.any():
        zero = torch.zeros((), device=x.device)
        arange = torch.arange(x.size(fill_index), device=x.device)
        for _ in range(x.dim() - fill_index if fill_index >= 0 else -fill_index):
            arange.unsqueeze_(-1)
        indices = torch.where(mask, zero, arange)
        _, index = indices.cummax(dim=fill_index)
        x = x.gather(dim=fill_index, index=index)
```

Reply:
I don't see how it will assuming it passes that test. 

Anyway, to avoid that altogether, I think this is a neater fix
```
    mask = torch.isnan(x)
    if mask.any():
        cumsum_mask = (~mask).cumsum(fill_index)
        cumsum_mask[mask] = 0
        _, index = (cumsum_mask).cummax(dim=fill_index)
        x = x.gather(dim=fill_index, index=index)
```

Reply:
On the first point: you're assuming that it does the same thing when the tensor has 4 elements as when the tensor has 234 elements, for example. It need not.

Agreed, that is a neater fix. I like it.